script/modelling/space_shuttle_reentry_trajectory.jl

```diff
@@ -0,0 +1,281 @@
+γₛ = deg2rad(-1)  # flight path angle (rad)
+ψₛ = deg2rad(90)  # azimuth (rad)
+αₛ = deg2rad(0)   # angle of attack (rad)
+βₛ = deg2rad(0)   # banck angle (rad)
+tₛ = 1.00         # time step (sec)
+
+# Final conditions, the so-called Terminal Area Energy Management (TAEM)
+hₜ = 0.8          # altitude (ft) / 1e5
+vₜ = 0.25         # velocity (ft/sec) / 1e4
+γₜ = deg2rad(-5)  # flight path angle (rad)
+
+# Fix initial conditions
+fix(scaled_h[1], hₛ; force=true)
+fix(       ϕ[1], ϕₛ; force=true)
+fix(       θ[1], θₛ; force=true)
+fix(scaled_v[1], vₛ; force=true)
+fix(       γ[1], γₛ; force=true)
+fix(       ψ[1], ψₛ; force=true)
+
+# Fix final conditions
+fix(scaled_h[n], hₜ; force=true)
+fix(scaled_v[n], vₜ; force=true)
+fix(       γ[n], γₜ; force=true)
+
+# Initial guess: linear interpolation between boundary conditions
+xₛ = [hₛ, ϕₛ, θₛ, vₛ, γₛ, ψₛ, αₛ, βₛ, tₛ]
+xₜ = [hₜ, ϕₛ, θₛ, vₜ, γₜ, ψₛ, αₛ, βₛ, tₛ]
+interp_linear = LinearInterpolation([1, n], [xₛ, xₜ])
+initial_guess = mapreduce(transpose, vcat, interp_linear.(1:n))
+set_start_value.(all_variables(model), vec(initial_guess))
+
+# Functions to restore `h` and `v` to their true scale
+@NLexpression(model, h[j = 1:n], scaled_h[j] * 1e5)
+@NLexpression(model, v[j = 1:n], scaled_v[j] * 1e4)
+
+# Helper functions
+# @NLexpression(model, c_L[j = 1:n], a₀ + a₁ * rad2deg(α[j]))
+# @NLexpression(model, c_D[j = 1:n], b₀ + b₁ * rad2deg(α[j]) + b₂ * rad2deg(α[j])^2)
+@NLexpression(model, c_L[j = 1:n], a₀ + a₁ * (180 * α[j] / π))
+@NLexpression(model, c_D[j = 1:n], b₀ + b₁ * (180 * α[j] / π) + b₂ * (180 * α[j] / π)^2)
+@NLexpression(model,   ρ[j = 1:n], ρ₀ * exp(-h[j] / hᵣ))
+@NLexpression(model,   D[j = 1:n], 0.5 * c_D[j] * S * ρ[j] * v[j]^2)
+@NLexpression(model,   L[j = 1:n], 0.5 * c_L[j] * S * ρ[j] * v[j]^2)
+@NLexpression(model,   r[j = 1:n], Rₑ + h[j])
+@NLexpression(model,   g[j = 1:n], μ / r[j]^2)
+
+# Motion of the vehicle as a differential-algebraic system of equations (DAEs)
+@NLexpression(model, δh[j = 1:n], v[j] * sin(γ[j]))
+@NLexpression(model, δϕ[j = 1:n], (v[j] / r[j]) * cos(γ[j]) * sin(ψ[j]) / cos(θ[j]))
+@NLexpression(model, δθ[j = 1:n], (v[j] / r[j]) * cos(γ[j]) * cos(ψ[j]))
+@NLexpression(model, δv[j = 1:n], -(D[j] / m) - g[j] * sin(γ[j]))
+@NLexpression(model, δγ[j = 1:n], (L[j] / (m * v[j])) * cos(β[j]) + cos(γ[j]) * ((v[j] / r[j]) - (g[j] / v[j])))
+@NLexpression(model, δψ[j = 1:n], (1 / (m * v[j] * cos(γ[j]))) * L[j] * sin(β[j]) + (v[j] / (r[j] * cos(θ[j]))) * cos(γ[j]) * sin(ψ[j]) * sin(θ[j]))
+
+# System dynamics
+for j in 2:n
+    i = j - 1  # index of previous knot
+
+    # Rectangular integration
+    @NLconstraint(model, h[j] == h[i] + Δt[i] * δh[i])
+    @NLconstraint(model, ϕ[j] == ϕ[i] + Δt[i] * δϕ[i])
+    @NLconstraint(model, θ[j] == θ[i] + Δt[i] * δθ[i])
+    @NLconstraint(model, v[j] == v[i] + Δt[i] * δv[i])
+    @NLconstraint(model, γ[j] == γ[i] + Δt[i] * δγ[i])
+    @NLconstraint(model, ψ[j] == ψ[i] + Δt[i] * δψ[i])
+
+    # # Trapezoidal integration
```

Comment:
Do you think we could include a new option to switch between rectangular and trapezoidal integration?

Reply:
Of course! What do you have in mind? Would something like this work?
```julia
# (...)

integration_rule = "rectangular"  # Possible values: rectangular, trapezoidal.

# (...)

# System dynamics
for j in 2:n
    i = j - 1  # index of previous knot

    if integration_rule == "rectangular"
        # Rectangular integration
        @NLconstraint(model, h[j] == h[i] + Δt[i] * δh[i])
        @NLconstraint(model, ϕ[j] == ϕ[i] + Δt[i] * δϕ[i])
        @NLconstraint(model, θ[j] == θ[i] + Δt[i] * δθ[i])
        @NLconstraint(model, v[j] == v[i] + Δt[i] * δv[i])
        @NLconstraint(model, γ[j] == γ[i] + Δt[i] * δγ[i])
        @NLconstraint(model, ψ[j] == ψ[i] + Δt[i] * δψ[i])
    elseif integration_rule == "trapezoidal"
        # Trapezoidal integration
        @NLconstraint(model, h[j] == h[i] + 0.5 * Δt[i] * (δh[j] + δh[i]))
        @NLconstraint(model, ϕ[j] == ϕ[i] + 0.5 * Δt[i] * (δϕ[j] + δϕ[i]))
        @NLconstraint(model, θ[j] == θ[i] + 0.5 * Δt[i] * (δθ[j] + δθ[i]))
        @NLconstraint(model, v[j] == v[i] + 0.5 * Δt[i] * (δv[j] + δv[i]))
        @NLconstraint(model, γ[j] == γ[i] + 0.5 * Δt[i] * (δγ[j] + δγ[i]))
        @NLconstraint(model, ψ[j] == ψ[i] + 0.5 * Δt[i] * (δψ[j] + δψ[i]))
    else
        @error "Unexpected integration rule '$(integration_rule)'"
    end
end

# (...)
```

Reply:
That would be perfect